client/stmt.go

```diff
@@ -275,14 +275,33 @@ func (c *Conn) Prepare(query string) (*Stmt, error) {
 	}
 
 	if s.params > 0 {
-		if err := s.conn.readUntilEOF(); err != nil {
-			return nil, errors.Trace(err)
+		for range s.params {
+			if _, err := s.conn.ReadPacket(); err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+		if s.conn.capability&mysql.CLIENT_DEPRECATE_EOF == 0 {
+			if packet, err := s.conn.ReadPacket(); err != nil {
+				return nil, errors.Trace(err)
+			} else if !c.isEOFPacket(packet) {
+				return nil, mysql.ErrMalformPacket
+			}
 		}
 	}
 
 	if s.columns > 0 {
-		if err := s.conn.readUntilEOF(); err != nil {
-			return nil, errors.Trace(err)
+		// TODO process when CLIENT_CACHE_METADATA enabled
```

Comment:
Do you mean `MARIADB_CLIENT_CACHE_METADATA`? https://mariadb.com/docs/server/reference/clientserver-protocol/mariadb-protocol-differences-with-mysql

I want to understand what's the effect after this PR. Hope it will not break existing program with that capabilities flag.

Reply:
We don't currently support those flags, so it won't affect. We have code that reads until eof to skip info that certain options would omit from protocol